pytket/pytket/rmbqc/mpattern.py

```diff
@@ -0,0 +1,533 @@
+        easiest way to split a zx diagram into multiple disjoint zx.diagrams is
+        to first identify X disjoint clusters of qubits, the copy the original
+        graph into X graphs, and then for each of the copies remove all the
+        vertices outside the corresponding cluster.
+        
+        :param g:       A pyzx graph representing a zx diagram.
+        :param type:    GraphS
+        
+        :param io_map:  A dictionary containing the current input/output mappings
+                            which map the input/output qubits of the original circuit
+                            to the corresponding vertices in the ZX diagram.
+        :param type:    Dict[str,Dict[Qubit,int]]
+        """
+        original_labels = sorted(
+            list(g.vertices())
+        )  # Sort the original vertices in ascending order.
+        # The following code will update the i/o map with the new labels of the
+        # input/output qubits after the squish.
+        for i in io_map["i"].keys():
+            for v in range(len(original_labels)):
+                if io_map["i"][i] == original_labels[v]:
+                    io_map["i"][i] = v
+                    break
+        for o in io_map["o"].keys():
+            for v in range(len(original_labels)):
+                if io_map["o"][o] == original_labels[v]:
+                    io_map["o"][o] = v
+                    break
+                
+    def remove_redundant(self, g: GraphS, io_map: Dict[str, Dict[Qubit, int]]) -> None:
+        """
+        Removes simples edges from a zx diagram by merging the connected
+        vertices.
+        
+        :param g:       A zx diagram with some remaining simple edges we want to remove.
+        :param type:    GraphS
+        
+        :param io_map:  A dictionary containing the current input/output mapping
+                            to map the input/output qubits of the original circuit
+                            to their corresponding vertices in the ZX diagram.
+        :param type:    Dict[str,Dict[Qubit,int]]
+        """
+        simple_edges = set()
+        for edge in g.edge_set():
+            if g.edge_type(edge) == 1:
+                simple_edges.add(edge)
+        for edge in simple_edges:
+            v1, v2 = edge
+            g.remove_edge(edge)
+            is_boundary = True
+            removing_input = True
+            remove_vertex = v1
+            keep_vertex = v2
+            if v1 in g.inputs:
+                remove_vertex = v1
+                keep_vertex = v2
+            elif v2 in g.inputs:
+                remove_vertex = v2
+                keep_vertex = v1
+            elif v1 in g.outputs:
+                removing_input = False
+            elif v2 in g.outputs:
+                removing_input = False
+                remove_vertex = v2
+                keep_vertex = v1
+            else:
+                is_boundary = False
+            g.set_row(keep_vertex, g.row(remove_vertex))
```

Comment:
mentioned above - this whole method probably needs an overhaul because those conditions can be removed - but it may mess with other stuff so not sure whether to do it right now